proposals/NNNN-fix-expressible-by-string-interpolation.md

```diff
@@ -0,0 +1,792 @@
+   `ExpressibleByStringLiteral`;
+   
+2. Introducing a `StringInterpolationSegment` generic enum and making 
+   `init(stringInterpolation:)` take a variadic list of them; and
+
+3. Removing the `init(stringInterpolationSegment:)` call.
+
+Combined with the new rule of interpreting interpolations as initializer 
+calls, the generated code for the `"Hello, \(name)!"` example above 
+becomes:
+
+```swift
+.init(stringInterpolation:
+	.literal("Hello, "),
+	.interpolation(.init(stringInterpolationSegment: name)),
+	.literal("!")
+)
+````
+
+This fixes both problems at once: literal and interpolated segments are 
+clearly labeled for the initializer to interpret, and the type can 
+tightly control the segment types using `StringLiteralType`, 
+`StringInterpolationType`, and the `StringInterpolationType`'s 
+initializers.
+
+## Detailed design
+
+A prototype of this design is available in [this pull request][enum-pr].
+Note that this prototype is not production-ready; it is more proof of 
+concept than final implementation.
+
+  [enum-pr]: https://github.com/apple/swift/pull/8352
+
+### New `ExpressibleByStringInterpolation` design
+
+The changes include:
+
+1. `ExpressibleByStringInterpolation` is made to refine 
+   `ExpressibleByStringLiteral`.
+
+2. A new unconstrained associated type, `StringInterpolationType`, 
+   is introduced.
+
+3. The `init(stringInterpolationSegment:)` initializer is removed.
+
+4. A new enum, `StringInterpolationSegment`, is introduced. It has 
+   two generic parameters, `Literal` and `Interpolation`, and two 
+   cases, `literal` and `interpolation`.
+
+5. The `init(stringInterpolation:)` initializer is modified so that, 
+   instead of having a variadic `Self` parameter, it has a variadic 
+   `StringInterpolationSegment<StringLiteralType, StringInterpolationType>`
+   parameter. (Passing both as separate generic parameters, rather 
+   than specifying `Self` as the generic parameter, means that 
+   defining `init(stringInterpolation:)` is enough to pin these types 
+   for the associated type inference engine.)
+
+6. The deprecation is dropped, and the underscored version becomes a 
+   typealias.
+
+7. An extension on `ExpressibleByStringInterpolation` is introduced 
+   which implements `init(stringLiteral:)` in terms of 
+   `init(stringInterpolation:)`.
+
+Exact listing:
+
+```swift
+/// A portion of a string literal containing interpolated expressions.
+/// 
+/// An interpolated string literal is divided into many segments. Each 
+/// is either a literal segment, containing a chunk of text found in 
+/// the source code, or an interpolated segment, containing the result 
+/// of an expression executed at runtime.
+/// 
+/// This type has two generic parameters. The first is the type of the 
+/// `literal` case's associated value; the second is the type of the 
+/// `interpolation` case's associated value. The `Literal` generic 
+/// parameter must be `ExpressibleByStringLiteral`.
```

Comment:
Yes, I suppose so. (For some reason, I thought that `ExpressibleByStringLiteral` refined its `_Builtin` equivalent.) It's sort of implied by the fact that `StringLiteralType` has that limitation, but no need to be coy about it.